precise/iaf_psc_exp_ps_lossless.cpp

```diff
@@ -0,0 +1,671 @@
+  {
+    throw BadProperty( "All time constants must be strictly positive." );
+  }
+
+  if ( tau_m_ == tau_ex_ || tau_m_ == tau_in_ )
+  {
+    throw BadProperty(
+      "Membrane and synapse time constant(s) must differ."
+      "See note in documentation." );
+  }
+
+  return delta_EL;
+}
+
+void
+nest::iaf_psc_exp_ps_lossless::State_::get( DictionaryDatum& d,
+  const Parameters_& p ) const
+{
+  def< double >( d, names::V_m, y2_ + p.E_L_ ); // Membrane potential
+  def< bool >( d, names::is_refractory, is_refractory_ );
+  def< double >(
+    d, names::t_spike, Time( Time::step( last_spike_step_ ) ).get_ms() );
+  def< double >( d, names::offset, last_spike_offset_ );
+  def< double >( d, names::I_syn_ex, I_syn_ex_ );
+  def< double >( d, names::I_syn_in, I_syn_in_ ); // y1 state
+  def< double >( d, names::I_syn, I_syn_ex_ + I_syn_in_ );
+}
+
+void
+nest::iaf_psc_exp_ps_lossless::State_::set( const DictionaryDatum& d,
+  const Parameters_& p,
+  double delta_EL )
+{
+  if ( updateValue< double >( d, names::V_m, y2_ ) )
+  {
+    y2_ -= p.E_L_;
+  }
+  else
+  {
+    y2_ -= delta_EL;
+  }
+
+  updateValue< double >( d, names::I_syn_ex, I_syn_ex_ );
+  updateValue< double >( d, names::I_syn_in, I_syn_in_ );
+}
+
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::iaf_psc_exp_ps_lossless::iaf_psc_exp_ps_lossless()
+  : Archiving_Node()
+  , P_()
+  , S_()
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+nest::iaf_psc_exp_ps_lossless::iaf_psc_exp_ps_lossless(
+  const iaf_psc_exp_ps_lossless& n )
+  : Archiving_Node( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::iaf_psc_exp_ps_lossless::init_state_( const Node& proto )
+{
+  const iaf_psc_exp_ps_lossless& pr =
+    downcast< iaf_psc_exp_ps_lossless >( proto );
+
+  S_ = pr.S_;
+}
+
+void
+nest::iaf_psc_exp_ps_lossless::init_buffers_()
+{
+  B_.events_.resize();
+  B_.events_.clear();
+  B_.currents_.clear(); // includes resize
+  B_.logger_.reset();
+}
+
+void
+nest::iaf_psc_exp_ps_lossless::calibrate()
+{
+  B_.logger_
+    .init(); // ensures initialization in case mm connected after Simulate
+
+  V_.h_ms_ = Time::get_resolution().get_ms();
+
+  V_.expm1_tau_m_ = numerics::expm1( -V_.h_ms_ / P_.tau_m_ );
+  V_.expm1_tau_ex_ = numerics::expm1( -V_.h_ms_ / P_.tau_ex_ );
+  V_.expm1_tau_in_ = numerics::expm1( -V_.h_ms_ / P_.tau_in_ );
+  V_.P20_ = -P_.tau_m_ / P_.c_m_ * V_.expm1_tau_m_;
+  V_.P21_ex_ = -P_.tau_m_ * P_.tau_ex_ / ( P_.tau_m_ - P_.tau_ex_ ) / P_.c_m_
+    * ( V_.expm1_tau_ex_ - V_.expm1_tau_m_ );
+  V_.P21_in_ = -P_.tau_m_ * P_.tau_in_ / ( P_.tau_m_ - P_.tau_in_ ) / P_.c_m_
+    * ( V_.expm1_tau_in_ - V_.expm1_tau_m_ );
```

Comment:
The expressions for `V_.P20_{ex, in}_` are a bit hard to read. I think you can multiply `-P_.tau_m_ * P_tau_{ex, in}_` by `P_.c_m_` because you are dividing twice.

Reply:
I think I will leave them as is, corresponding expressions look the same in all other models. Once we generate code from NESTML it will look differently anyways.